### PR TITLE
Support FIPO

### PR DESCRIPTION
**Pull request by li2zhi**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [x] Document Updates
- [ ] More Models or Datasets Support

# PR information

FIPO is a value-free RL algorithm proposed in [FIPO: Future-guided Importance Policy Optimization](https://arxiv.org/abs/2603.19835). It is designed to improve long-form reasoning training by refining token-level credit assignment in GRPO/DAPO-style optimization.

## Algorithm Overview

FIPO keeps the standard DAPO scaffold, but changes how token-level updates are weighted. The local signal is the signed log-probability shift between the current and old policy:

$$
\Delta \log p_t = \log \pi_\theta(y_t \mid x, y_{1:t-1}) - \log \pi_{old}(y_t \mid x, y_{1:t-1})
$$

Positive values mean the token is being reinforced, while negative values mean it is being suppressed. Since reasoning is sequential, FIPO then accumulates this signal over the future trajectory:

$$
FutureKL_t = \sum_{k=t}^{T} M_k \cdot \gamma^{k-t} \cdot \Delta \log p_k
$$

Positive `FutureKL_t` means the future following token `t` is being reinforced; negative `FutureKL_t` means it is being suppressed. The decay window keeps the signal local enough to stay stable, while the mask removes extreme-ratio outliers.

FIPO maps this future signal into a bounded influence weight:

$$
f_t = clip(\exp(FutureKL_t), 1-\epsilon_{f,low}, 1+\epsilon_{f,high}), \quad \tilde{A}_t = \hat{A}_t \cdot f_t
$$

```text
FutureKL_t = discounted_sum_of_future_logprob_shifts(t)
weighted_advantage_t = A_t * clip(exp(FutureKL_t), low, high)
```

The final token-level FIPO loss keeps the standard clipped PPO/DAPO form, but replaces the original advantage with the future-aware one:

$$
r_t = \frac{\pi_\theta(y_t \mid x, y_{1:t-1})}{\pi_{old}(y_t \mid x, y_{1:t-1})}
$$

$$
L_t^{FIPO} = min(r_t \tilde{A}_t,\; clip(r_t, 1-\epsilon, 1+\epsilon)\tilde{A}_t)
$$

Tokens that lead into preferred futures are **amplified**, while tokens that lead into suppressed futures are **attenuated**. Clipping keeps this modulation stable. The final DAPO-style loss therefore stays clipped and simple, but the advantage term becomes **future-aware** rather than uniformly inherited from the final outcome.

## Reference

Paper: [FIPO: Future-guided Importance Policy Optimization](https://arxiv.org/abs/2603.19835)

### CODE REVIEW

**Comment by hjh0119** [comment]
thanks for your contribution

is this pr ready to be merged?

**Comment by li2zhi** [comment]
Yes. Please let me know if there are any remaining concerns or changes needed.

File: docs/source/Instruction/GRPO/AdvancedResearch/FIPO.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+| `--loss_type`             | `str`   | `grpo` | 设置为`fipo` 启用 FIPO loss                                                                                         |
+| `--delta`                 | `float` | `None` | 启用后会同时用于 Future-KL 高 IS ratio token 过滤和主 loss 的 dual-clip 上限，应大于 `1 + epsilon_high`，对齐FIPO 32B训练脚本建议设置为 `10.0` |
+| `--fipo_decay_rate`       | `float` | `32.0` | Future-KL 折扣半衰参数，实际折扣为`2 ** (-1 / fipo_decay_rate)`                                                            |
+| `--fipo_clip_range`       | `float` | `0.2`  | influence weight 裁剪范围；`0.2` 表示默认裁剪到 `[0.8, 1.2]`                                                               |
+| `--fipo_clip_high_only`   | `bool`  | `true` | 若为`true`，权重只裁剪到 `[1.0, 1.0 + fipo_clip_range]`，更偏向放大正 Future-KL                                                |
+| `--fipo_detach_weight`    | `bool`  | `true` | 是否对 influence weight 截断梯度                                                                                      |
+| `--fipo_safety_threshold` | `float` | `4.0`  | 负 advantage 且 IS ratio 超过该阈值时，将 FIPO 权重限制到 `[0.8, 1.0]` 以避免过度惩罚                                                |
```
Comment (hjh0119) [inline docs/source/Instruction/GRPO/AdvancedResearch/FIPO.md:48]: Could you also add the corresponding documentation in both the loss_types doc and the Command-line-parameters doc?

File: examples/train/grpo/internal/fipo.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    --attn_impl flash_attn \
+    --beta 0 \
+    --dynamic_sample true \
+    --loss_type fipo \
+    --delta 10.0 \
+    --epsilon_high 0.28 \
+    --fipo_decay_rate 32 \
+    --fipo_clip_range 0.2 \
+    --fipo_clip_high_only true \
+    --fipo_detach_weight true \
```
Comment (hjh0119) [inline examples/train/grpo/internal/fipo.sh]: It would be better if it aligns with the experiments in the paper (Appendix B)

File: docs/source/Instruction/GRPO/AdvancedResearch/FIPO.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+|---------------------------|---------|--------|----------------------------------------------------------------------------------------------------------------|
+| `--loss_type`             | `str`   | `grpo` | 设置为`fipo` 启用 FIPO loss                                                                                         |
+| `--delta`                 | `float` | `None` | 启用后会同时用于 Future-KL 高 IS ratio token 过滤和主 loss 的 dual-clip 上限，应大于 `1 + epsilon_high`，对齐FIPO 32B训练脚本建议设置为 `10.0` |
+| `--fipo_decay_rate`       | `float` | `32.0` | Future-KL 折扣半衰参数，实际折扣为`2 ** (-1 / fipo_decay_rate)`                                                            |
+| `--fipo_clip_range`       | `float` | `0.2`  | influence weight 裁剪范围；`0.2` 表示默认裁剪到 `[0.8, 1.2]`                                                               |
+| `--fipo_clip_high_only`   | `bool`  | `true` | 若为`true`，权重只裁剪到 `[1.0, 1.0 + fipo_clip_range]`，更偏向放大正 Future-KL                                                |
+| `--fipo_detach_weight`    | `bool`  | `true` | 是否对 influence weight 截断梯度                                                                                      |
```
Comment (hjh0119) [inline docs/source/Instruction/GRPO/AdvancedResearch/FIPO.md]: `fipo_detach_weight` should be removed and become the default behavior, 
aligning with the official implementation.

File: swift/megatron/trainers/grpo_trainer.py [hunk 4]
```diff
@@ -484,12 +491,53 @@ def _generate_and_score_completions(self, batch):
...
+            high_ratio_mask = coef_1 > delta
+            future_kl_delta = torch.where(high_ratio_mask, torch.zeros_like(future_kl_delta), future_kl_delta)
+
+        future_kl = torch.zeros_like(future_kl_delta)
+        running = torch.zeros_like(future_kl_delta[:, 0])
+        gamma = torch.as_tensor(self.fipo_gamma, dtype=log_ratio.dtype, device=log_ratio.device)
+        for token_idx in range(future_kl_delta.shape[1] - 1, -1, -1):
+            valid_token = completion_mask[:, token_idx]
+            running = torch.where(valid_token, future_kl_delta[:, token_idx] + gamma * running, running)
+            future_kl[:, token_idx] = torch.where(valid_token, running, 0.0)
```
Comment (hjh0119) [inline swift/megatron/trainers/grpo_trainer.py]: The per-token loop may be inefficient. We could consider using 
chunked matrix multiplication, following the paper / official implementation.

**Review by hjh0119 — APPROVED** [review]

**Comment by hjh0119** [comment]
Thanks for your contribution! I've left a few comments

**Comment by li2zhi** [comment]
Thanks for your review!

I've addressed the comments and pushed the updates. Please let me know if anything else needs to be changed.

**Comment by hjh0119** [comment]
LGTM thanks!
